### ISSUE TITLE
Feature Request: VIM keymap

### ISSUE THREAD

**Issue by carsondoesbusiness** [heyman/heynote#24]
Please include VIM key mapping!

**Comment by nilsherzig** [comment]
coremirror has a vim mode https://codemirror.net/5/demo/vim.html should be "easy" to implement

**Comment by heyman** [comment]
I'd like to have this. Implementing it should probably start with some refactoring. My Emacs keymap (it's actually just the subset of the Emacs keybindings I happen to be using) is currently a bit of a duct tape implementation. Commands should be more formalized into an internal API, and once that is in place it would be nice to have a proper UI for customizing the mapped keys.

There are implementation details that will make it non-trivial. E.g. we can't trigger copy/cut/paste commands from within the renderer so there's currently code in the Electron main process that listens for those keys (https://github.com/heyman/heynote/blob/main/electron/keymap.js). Should be solvable though.

**Comment by rdslw** [comment]
Would love to see it :)

**Comment by tarasowski** [comment]
bump!

**Comment by brendanbond** [comment]
From the draft PR:
> It's likely that we need to copy these implementations into Heynote and make them block aware for the Vim mode.

It would be nice to simply "extend" the operators, so I've [asked about a potential API change to the vim dependency](https://github.com/replit/codemirror-vim/issues/155) that will hopefully allow us to make the operator aware of block state. @heyman let me know if you'd be open to this type of implementation.

**Comment by brendanbond** [comment]
@heyman yeah that's kind of my thinking - if we find ourselves in "the middle" of a block separator (i.e. "\n<separator>") we might already know that we need to offset the entire separator

in any case, thanks for the insight! hopefully can get some time this weekend to tinker with it

**Comment by rdslw** [comment]
any estimates on the progress @heyman ?

**Comment by heyman** [comment]
Copying this from [my comment](https://github.com/heyman/heynote/pull/51#issuecomment-1889484295) on the Draft Vim PR:

> I'd love to merge support for Vim key bindings, but then someone needs to do the work. And unfortunately, it's unlikely to be me since I've never used Vim and need to google how to quit anytime Vi opens up in a terminal.

**Comment by TangYuanLonelyCat** [comment]
> I'd love to merge support for Vim key bindings, but then someone needs to do the work. And unfortunately, it's unlikely to be me since I've never used Vim and need to google how to quit anytime Vi opens up in a terminal.

I'm trying to solve this issue, but it looks quite **complicated**.

**Comment by hyperkubo** [comment]
Some of us really need this feature. It'll be awesome.
